src/thread_task.rs

```diff
@@ -813,869 +813,746 @@ unsafe fn delayed_fg_task(c: *const Rav1dContext, ttd: *mut TaskThreadData) {
-                    let mut i: c_uint = 0 as c_int as c_uint;
-                    loop {
-                        if !(i < (*c).n_fc) {
-                            current_block = 5601891728916014340;
-                            break;
+    'outer: while !(*tc).task_thread.die {
+        if ::core::intrinsics::atomic_load_seqcst((*c).flush) != 0 {
+            park(tc, ttd);
+            continue;
+        }
+
+        merge_pending(c);
+        if (*ttd).delayed_fg.exec != 0 {
+            // run delayed film grain first
+            delayed_fg_task(c, ttd);
+            continue;
+        }
+
+        let mut found = false;
+        let mut f = 0 as *mut Rav1dFrameContext;
+        let mut t = 0 as *mut Rav1dTask;
+        let mut prev_t = 0 as *mut Rav1dTask;
+        if (*c).n_fc > 1 as c_uint {
+            // run init tasks second
+            for i in 0..(*c).n_fc {
+                let first: c_uint = ::core::intrinsics::atomic_load_seqcst(&mut (*ttd).first);
+                f = &mut *((*c).fc).offset(first.wrapping_add(i).wrapping_rem((*c).n_fc) as isize)
+                    as *mut Rav1dFrameContext;
+                if ::core::intrinsics::atomic_load_seqcst(
+                    &mut (*f).task_thread.init_done as *mut atomic_int,
+                ) != 0
+                {
+                    continue;
+                }
+                t = (*f).task_thread.task_head;
+                if t.is_null() {
+                    continue;
+                }
+                if (*t).type_0 as c_uint == RAV1D_TASK_TYPE_INIT as c_int as c_uint {
+                    found = true;
+                    break;
+                }
+                if (*t).type_0 as c_uint == RAV1D_TASK_TYPE_INIT_CDF as c_int as c_uint {
+                    // XXX This can be a simple else, if adding tasks of both
+                    // passes at once (in dav1d_task_create_tile_sbrow).
+                    // Adding the tasks to the pending Q can result in a
+                    // thread merging them before setting init_done.
+                    // We will need to set init_done before adding to the
+                    // pending Q, so maybe return the tasks, set init_done,
+                    // and add to pending Q only then.
+                    let p1 = (if !((*f).in_cdf.progress).is_null() {
+                        ::core::intrinsics::atomic_load_seqcst((*f).in_cdf.progress)
+                    } else {
+                        1 as c_int as c_uint
+                    }) as c_int;
+                    if p1 != 0 {
+                        ::core::intrinsics::atomic_or_seqcst(
+                            &mut (*f).task_thread.error,
+                            (p1 == TILE_ERROR) as c_int,
+                        );
+                        found = true;
+                        break;
+                    }
+                }
+            }
+        }
+        if !found {
+            'decoding: while (*ttd).cur < (*c).n_fc {
```

Comment:
If you add back some of the comments in a `fn`, type, etc., could you add back all of them?  E.x. here there's a `// run decoding tasks last`.  That way it's easy to tell if all the comments have been added back or if we should look in the C for them.

Reply:
Meant to, missed that one.